From: Hannes Reinecke <hare@suse.de>
Date: Thu, 10 Jan 2013 14:22:25 +0100
Subject: dasd: make DASD_FLAG_TIMEOUT setting more robust
Patch-Mainline: n/a
References: FATE#311379

We should only start aborting requests if we could set
the flag. And we should print out a message if the
flag had been modified.

Signed-off-by: Hannes Reinecke <hare@suse.de>
---
 drivers/s390/block/dasd_ioctl.c |   12 ++++++++----
 1 files changed, 8 insertions(+), 4 deletions(-)

diff --git a/drivers/s390/block/dasd_ioctl.c b/drivers/s390/block/dasd_ioctl.c
index f757e4e..92b8a55 100644
--- a/drivers/s390/block/dasd_ioctl.c
+++ b/drivers/s390/block/dasd_ioctl.c
@@ -154,8 +154,12 @@ static int dasd_ioctl_timeout(struct dasd_block *block)
 	if (!capable (CAP_SYS_ADMIN))
 		return -EACCES;
 
+	if (test_and_set_bit(DASD_FLAG_TIMEOUT, &base->flags))
+		return 0;
+
+	dev_err(&base->cdev->dev, "ioctl timeout flag set\n");
+
 	spin_lock_irqsave(&block->request_queue_lock, flags);
-	set_bit(DASD_FLAG_TIMEOUT, &base->flags);
 	/*
 	 * We need to abort the first noretry requests as
 	 * there might be requests on the ccw_queue.
@@ -195,9 +199,9 @@ static int dasd_ioctl_resync(struct dasd_block *block)
 	if (!capable (CAP_SYS_ADMIN))
 		return -EACCES;
 
-	spin_lock_irqsave(get_ccwdev_lock(base->cdev), flags);
-	clear_bit(DASD_FLAG_TIMEOUT, &base->flags);
-	spin_unlock_irqrestore(get_ccwdev_lock(base->cdev), flags);
+	if (test_and_clear_bit(DASD_FLAG_TIMEOUT, &base->flags))
+		dev_err(&base->cdev->dev, "ioctl timeout flag unset\n");
+
 	return 0;
 }
 
-- 
1.7.4.2

